File: test/firestore_security_rules_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:fake_cloud_firestore/fake_cloud_firestore.dart';
import 'package:firebase_auth_mocks/firebase_auth_mocks.dart';
import 'package:cloud_firestore/cloud_firestore.dart';

/// Test suite for Firestore security rules
///
/// Note: These tests simulate the security rule logic since we cannot
/// directly test Firestore security rules in unit tests. For full
/// security rule testing, use the Firebase emulator with @firebase/rules-unit-testing
void main() {
  group('Firestore Security Rules Logic Tests', () {
    late FakeFirebaseFirestore firestore;
    late MockFirebaseAuth auth;

    setUp(() {
      firestore = FakeFirebaseFirestore();
      auth = MockFirebaseAuth();
    });

    group('Friend Request Security Logic', () {
      test(
        'should allow user to read their own sent friend requests',
        () async {
          // Create a friend request document
          await firestore.collection('friendRequests').doc('request1').set({
            'senderId': 'user1',
            'receiverId': 'user2',
            'senderUsername': 'testuser1',
            'status': 'pending',
            'createdAt': FieldValue.serverTimestamp(),
          });

          // User should be able to read their own sent request
          final doc = await firestore
              .collection('friendRequests')
              .doc('request1')
              .get();
          expect(doc.exists, isTrue);
          expect(doc.data()!['senderId'], equals('user1'));
        },
      );

      test(
        'should allow user to read their own received friend requests',
        () async {
          // Create a friend request where user is receiver
          await firestore.collection('friendRequests').doc('request2').set({
            'senderId': 'user2',
            'receiverId': 'user1',
            'senderUsername': 'testuser2',
            'status': 'pending',
            'createdAt': FieldValue.serverTimestamp(),
          });

          final doc = await firestore
              .collection('friendRequests')
              .doc('request2')
              .get();
          expect(doc.exists, isTrue);
          expect(doc.data()!['receiverId'], equals('user1'));
        },
      );

      test('should validate friend request creation data', () async {
        // Test valid friend request creation
        final validRequest = {
          'senderId': 'user1',
          'receiverId': 'user2',
          'senderUsername': 'testuser1',
          'status': 'pending',
          'createdAt': FieldValue.serverTimestamp(),
        };

        // Should succeed with valid data
        await firestore.collection('friendRequests').add(validRequest);

        // Validate required fields are present
        expect(validRequest.containsKey('senderId'), isTrue);
        expect(validRequest.containsKey('receiverId'), isTrue);
        expect(validRequest.containsKey('senderUsername'), isTrue);
        expect(validRequest.containsKey('status'), isTrue);
        expect(validRequest.containsKey('createdAt'), isTrue);
      });

      test('should prevent self friend requests', () {
        // Simulate validation logic that would be in security rules
        final requestData = {
          'senderId': 'user1',
          'receiverId': 'user1', // Same as sender
          'senderUsername': 'testuser1',
          'status': 'pending',
          'createdAt': FieldValue.serverTimestamp(),
        };

        // This should fail validation (senderId == receiverId)
        expect(requestData['senderId'], equals(requestData['receiverId']));

        // In real security rules, this would be prevented
        final isSelfRequest =
            requestData['senderId'] == requestData['receiverId'];
        expect(
          isSelfRequest,
          isTrue,
        ); // This validates the logic that should be prevented
      });
    });

    group('Friendship Security Logic', () {
      test('should allow users to read their own friendships', () async {
        await firestore.collection('friendships').doc('friendship1').set({
          'userId1': 'user1',
          'userId2': 'user2',
          'createdAt': FieldValue.serverTimestamp(),
        });

        final doc = await firestore
            .collection('friendships')
            .doc('friendship1')
            .get();
        expect(doc.exists, isTrue);

        // Both users should be able to read this friendship
        final data = doc.data()!;
        expect(data['userId1'], equals('user1'));
        expect(data['userId2'], equals('user2'));
      });

      test('should allow users to delete their own friendships', () async {
        await firestore.collection('friendships').doc('friendship2').set({
          'userId1': 'user1',
          'userId2': 'user2',
          'createdAt': FieldValue.serverTimestamp(),
        });

        // User should be able to delete friendship they're part of
        await firestore.collection('friendships').doc('friendship2').delete();

        final doc = await firestore
            .collection('friendships')
            .doc('friendship2')
            .get();
        expect(doc.exists, isFalse);
      });
    });

    group('Folder Security Logic', () {
      test('should allow owner to read/write their folders', () async {
        await firestore.collection('folders').doc('folder1').set({
          'name': 'Test Folder',
          'userId': 'user1',
          'createdAt': FieldValue.serverTimestamp(),
          'isShared': false,
          'isPublic': false,
          'isLocked': false,
          'contributorIds': [],
        });

        final doc = await firestore.collection('folders').doc('folder1').get();
        expect(doc.exists, isTrue);
        expect(doc.data()!['userId'], equals('user1'));
      });

      test('should allow contributors to access shared folders', () async {
        await firestore.collection('folders').doc('folder2').set({
          'name': 'Shared Folder',
          'userId': 'user1',
          'createdAt': FieldValue.serverTimestamp(),
          'isShared': true,
          'isPublic': false,
          'isLocked': false,
          'contributorIds': ['user2', 'user3'],
        });

        final doc = await firestore.collection('folders').doc('folder2').get();
        final data = doc.data()!;

        // Contributors should be in the contributorIds list
        expect(data['contributorIds'], contains('user2'));
        expect(data['contributorIds'], contains('user3'));
      });

      test('should allow anyone to read public folders', () async {
        await firestore.collection('folders').doc('folder3').set({
          'name': 'Public Folder',
          'userId': 'user1',
          'createdAt': FieldValue.serverTimestamp(),
          'isShared': false,
          'isPublic': true,
          'isLocked': false,
          'contributorIds': [],
        });

        final doc = await firestore.collection('folders').doc('folder3').get();
        expect(doc.data()!['isPublic'], isTrue);
      });

      test('should prevent contributors from modifying locked folders', () {
        final folderData = {
          'name': 'Locked Folder',
          'userId': 'user1',
          'isShared': true,
          'isLocked': true, // Folder is locked
          'contributorIds': ['user2'],
        };

        // Simulate security rule logic
        final contributorIds = folderData['contributorIds'] as List<dynamic>;
        final isContributor = contributorIds.contains('user2');
        final isLocked = folderData['isLocked'] as bool;
        final canModify = !isLocked || folderData['userId'] == 'user2';

        expect(isContributor, isTrue);
        expect(isLocked, isTrue);
        expect(canModify, isFalse); // Contributor cannot modify locked folder
      });
    });

    group('Scheduled Message Security Logic', () {
      test('should allow sender and recipient to read messages', () async {
        await firestore.collection('scheduledMessages').doc('message1').set({
          'senderId': 'user1',
          'recipientId': 'user2',
          'textContent': 'Hello future!',
          'scheduledFor': Timestamp.fromDate(
            DateTime.now().add(Duration(days: 1)),
          ),
          'createdAt': FieldValue.serverTimestamp(),
          'status': 'pending',
        });

        final doc = await firestore
            .collection('scheduledMessages')
            .doc('message1')
            .get();
        final data = doc.data()!;

        // Both sender and recipient should be able to read
        expect(data['senderId'], equals('user1'));
        expect(data['recipientId'], equals('user2'));
      });

      test('should validate scheduled message creation', () {
        final now = DateTime.now();
        final futureDate = now.add(Duration(days: 1));
        final pastDate = now.subtract(Duration(days: 1));

        // Valid message data
        final validMessage = {
          'senderId': 'user1',
          'recipientId': 'user2',
          'textContent': 'Valid message',
          'scheduledFor': Timestamp.fromDate(futureDate),
          'createdAt': FieldValue.serverTimestamp(),
          'status': 'pending',
        };

        // Invalid message (scheduled for past)
        final invalidMessage = {
          'senderId': 'user1',
          'recipientId': 'user2',
          'textContent': 'Invalid message',
          'scheduledFor': Timestamp.fromDate(pastDate),
          'createdAt': FieldValue.serverTimestamp(),
          'status': 'pending',
        };

        // Simulate validation logic
        final validTimestamp = validMessage['scheduledFor'] as Timestamp;
        final invalidTimestamp = invalidMessage['scheduledFor'] as Timestamp;
        expect(
          validTimestamp.seconds > now.millisecondsSinceEpoch / 1000,
          isTrue,
        );
        expect(
          invalidTimestamp.seconds < now.millisecondsSinceEpoch / 1000,
          isTrue,
        );
      });

      test('should validate text content length', () {
        final shortText = 'Short message';
        final longText = 'x' * 6000; // Exceeds 5KB limit
        final validText = 'x' * 1000; // Within limit

        expect(shortText.isNotEmpty, isTrue);
        expect(longText.length > 5000, isTrue); // Should fail
        expect(validText.length <= 5000, isTrue); // Should pass
      });

      test('should allow sender to cancel pending messages', () {
        final messageData = {
          'senderId': 'user1',
          'recipientId': 'user2',
          'status': 'pending',
        };

        // Simulate cancellation logic
        final canCancel =
            messageData['senderId'] == 'user1' &&
            messageData['status'] == 'pending';

        expect(canCancel, isTrue);
      });
    });

    group('Username Validation Logic', () {
      test('should validate username format', () {
        final validUsernames = ['user123', 'test_user', 'User_Name_123'];
        final invalidUsernames = ['us', 'user@name', 'user name', 'x' * 31];

        for (final username in validUsernames) {
          expect(
            isValidUsername(username),
            isTrue,
            reason: 'Username $username should be valid',
          );
        }

        for (final username in invalidUsernames) {
          expect(
            isValidUsername(username),
            isFalse,
            reason: 'Username $username should be invalid',
          );
        }
      });
    });
  });
}

/// Helper function to validate username format
/// Mirrors the validation logic in Firestore security rules
bool isValidUsername(String username) {
  if (username.length < 3 || username.length > 30) return false;
  return RegExp(r'^[a-zA-Z0-9_]+$').hasMatch(username);
}

File: test/firestore-security-rules.test.js
const { initializeTestEnvironment, assertFails, assertSucceeds } = require('@firebase/rules-unit-testing');
const { doc, getDoc, setDoc, updateDoc, deleteDoc, collection, addDoc, query, where, getDocs } = require('firebase/firestore');

/**
 * Comprehensive test suite for Firestore security rules
 * 
 * To run these tests:
 * 1. Install dependencies: npm install --save-dev @firebase/rules-unit-testing
 * 2. Start Firebase emulator: firebase emulators:start --only firestore
 * 3. Run tests: npm test
 */

let testEnv;

beforeAll(async () => {
  testEnv = await initializeTestEnvironment({
    projectId: 'time-capsule-test',
    firestore: {
      rules: require('fs').readFileSync('firestore.rules', 'utf8'),
    },
  });
});

afterAll(async () => {
  await testEnv.cleanup();
});

beforeEach(async () => {
  await testEnv.clearFirestore();
});

describe('Friend Request Security Rules', () => {
  test('should allow user to read their own sent friend requests', async () => {
    const alice = testEnv.authenticatedContext('alice');
    const friendRequestRef = doc(alice.firestore(), 'friendRequests', 'request1');
    
    await assertSucceeds(setDoc(friendRequestRef, {
      senderId: 'alice',
      receiverId: 'bob',
      senderUsername: 'alice_user',
      status: 'pending',
      createdAt: new Date(),
    }));
    
    await assertSucceeds(getDoc(friendRequestRef));
  });
  
  test('should allow user to read their own received friend requests', async () => {
    const alice = testEnv.authenticatedContext('alice');
    const bob = testEnv.authenticatedContext('bob');
    const friendRequestRef = doc(alice.firestore(), 'friendRequests', 'request2');
    
    // Alice creates a friend request to Bob
    await assertSucceeds(setDoc(friendRequestRef, {
      senderId: 'alice',
      receiverId: 'bob',
      senderUsername: 'alice_user',
      status: 'pending',
      createdAt: new Date(),
    }));
    
    // Bob should be able to read the request sent to him
    await assertSucceeds(getDoc(doc(bob.firestore(), 'friendRequests', 'request2')));
  });
  
  test('should deny reading friend requests not involving the user', async () => {
    const alice = testEnv.authenticatedContext('alice');
    const charlie = testEnv.authenticatedContext('charlie');
    const friendRequestRef = doc(alice.firestore(), 'friendRequests', 'request3');
    
    await assertSucceeds(setDoc(friendRequestRef, {
      senderId: 'alice',
      receiverId: 'bob',
      senderUsername: 'alice_user',
      status: 'pending',
      createdAt: new Date(),
    }));
    
    // Charlie should not be able to read Alice's request to Bob
    await assertFails(getDoc(doc(charlie.firestore(), 'friendRequests', 'request3')));
  });
  
  test('should prevent self friend requests', async () => {
    const alice = testEnv.authenticatedContext('alice');
    const friendRequestRef = doc(alice.firestore(), 'friendRequests', 'selfRequest');
    
    await assertFails(setDoc(friendRequestRef, {
      senderId: 'alice',
      receiverId: 'alice', // Same as sender
      senderUsername: 'alice_user',
      status: 'pending',
      createdAt: new Date(),
    }));
  });
  
  test('should allow receiver to accept friend request', async () => {
    const alice = testEnv.authenticatedContext('alice');
    const bob = testEnv.authenticatedContext('bob');
    const friendRequestRef = doc(alice.firestore(), 'friendRequests', 'request4');
    
    // Alice sends friend request to Bob
    await assertSucceeds(setDoc(friendRequestRef, {
      senderId: 'alice',
      receiverId: 'bob',
      senderUsername: 'alice_user',
      status: 'pending',
      createdAt: new Date(),
    }));
    
    // Bob accepts the request
    await assertSucceeds(updateDoc(doc(bob.firestore(), 'friendRequests', 'request4'), {
      status: 'accepted',
      respondedAt: new Date(),
    }));
  });
  
  test('should deny sender from accepting their own request', async () => {
    const alice = testEnv.authenticatedContext('alice');
    const friendRequestRef = doc(alice.firestore(), 'friendRequests', 'request5');
    
    await assertSucceeds(setDoc(friendRequestRef, {
      senderId: 'alice',
      receiverId: 'bob',
      senderUsername: 'alice_user',
      status: 'pending',
      createdAt: new Date(),
    }));
    
    // Alice should not be able to accept her own request
    await assertFails(updateDoc(friendRequestRef, {
      status: 'accepted',
      respondedAt: new Date(),
    }));
  });
});

describe('Friendship Security Rules', () => {
  test('should allow users to read their own friendships', async () => {
    const alice = testEnv.authenticatedContext('alice');
    const friendshipRef = doc(alice.firestore(), 'friendships', 'friendship1');
    
    // Simulate system creating friendship (this would normally be done by Cloud Function)
    const admin = testEnv.authenticatedContext('admin', { admin: true });
    await assertSucceeds(setDoc(doc(admin.firestore(), 'friendships', 'friendship1'), {
      userId1: 'alice',
      userId2: 'bob',
      createdAt: new Date(),
    }));
    
    // Alice should be able to read the friendship
    await assertSucceeds(getDoc(friendshipRef));
  });
  
  test('should allow users to delete their own friendships', async () => {
    const alice = testEnv.authenticatedContext('alice');
    const admin = testEnv.authenticatedContext('admin', { admin: true });
    
    // Create friendship
    await assertSucceeds(setDoc(doc(admin.firestore(), 'friendships', 'friendship2'), {
      userId1: 'alice',
      userId2: 'bob',
      createdAt: new Date(),
    }));
    
    // Alice should be able to delete the friendship
    await assertSucceeds(deleteDoc(doc(alice.firestore(), 'friendships', 'friendship2')));
  });
  
  test('should deny users from creating friendships directly', async () => {
    const alice = testEnv.authenticatedContext('alice');
    const friendshipRef = doc(alice.firestore(), 'friendships', 'directFriendship');
    
    await assertFails(setDoc(friendshipRef, {
      userId1: 'alice',
      userId2: 'bob',
      createdAt: new Date(),
    }));
  });
});

describe('Folder Security Rules', () => {
  test('should allow owner to create and access their folders', async () => {
    const alice = testEnv.authenticatedContext('alice');
    const folderRef = doc(alice.firestore(), 'folders', 'aliceFolder');
    
    await assertSucceeds(setDoc(folderRef, {
      name: 'Alice\'s Folder',
      userId: 'alice',
      createdAt: new Date(),
      isShared: false,
      isPublic: false,
      isLocked: false,
      contributorIds: [],
    }));
    
    await assertSucceeds(getDoc(folderRef));
  });
  
  test('should allow contributors to access shared folders', async () => {
    const alice = testEnv.authenticatedContext('alice');
    const bob = testEnv.authenticatedContext('bob');
    const folderRef = doc(alice.firestore(), 'folders', 'sharedFolder');
    
    // Alice creates shared folder with Bob as contributor
    await assertSucceeds(setDoc(folderRef, {
      name: 'Shared Folder',
      userId: 'alice',
      createdAt: new Date(),
      isShared: true,
      isPublic: false,
      isLocked: false,
      contributorIds: ['bob'],
    }));
    
    // Bob should be able to read the shared folder
    await assertSucceeds(getDoc(doc(bob.firestore(), 'folders', 'sharedFolder')));
  });
  
  test('should allow anyone to read public folders', async () => {
    const alice = testEnv.authenticatedContext('alice');
    const charlie = testEnv.authenticatedContext('charlie');
    const folderRef = doc(alice.firestore(), 'folders', 'publicFolder');
    
    // Alice creates public folder
    await assertSucceeds(setDoc(folderRef, {
      name: 'Public Folder',
      userId: 'alice',
      createdAt: new Date(),
      isShared: false,
      isPublic: true,
      isLocked: false,
      contributorIds: [],
    }));
    
    // Charlie (not owner or contributor) should be able to read public folder
    await assertSucceeds(getDoc(doc(charlie.firestore(), 'folders', 'publicFolder')));
  });
  
  test('should deny access to private folders for non-owners/contributors', async () => {
    const alice = testEnv.authenticatedContext('alice');
    const charlie = testEnv.authenticatedContext('charlie');
    const folderRef = doc(alice.firestore(), 'folders', 'privateFolder');
    
    // Alice creates private folder
    await assertSucceeds(setDoc(folderRef, {
      name: 'Private Folder',
      userId: 'alice',
      createdAt: new Date(),
      isShared: false,
      isPublic: false,
      isLocked: false,
      contributorIds: [],
    }));
    
    // Charlie should not be able to read Alice's private folder
    await assertFails(getDoc(doc(charlie.firestore(), 'folders', 'privateFolder')));
  });
  
  test('should prevent contributors from modifying locked folders', async () => {
    const alice = testEnv.authenticatedContext('alice');
    const bob = testEnv.authenticatedContext('bob');
    const folderRef = doc(alice.firestore(), 'folders', 'lockedFolder');
    
    // Alice creates shared but locked folder
    await assertSucceeds(setDoc(folderRef, {
      name: 'Locked Folder',
      userId: 'alice',
      createdAt: new Date(),
      isShared: true,
      isPublic: false,
      isLocked: true,
      contributorIds: ['bob'],
    }));
    
    // Bob should not be able to modify locked folder
    await assertFails(updateDoc(doc(bob.firestore(), 'folders', 'lockedFolder'), {
      name: 'Modified Name',
    }));
  });
});

describe('Scheduled Message Security Rules', () => {
  test('should allow user to create scheduled message to friend', async () => {
    const alice = testEnv.authenticatedContext('alice');
    const messageRef = doc(alice.firestore(), 'scheduledMessages', 'message1');
    
    const futureDate = new Date();
    futureDate.setDate(futureDate.getDate() + 1);
    
    await assertSucceeds(setDoc(messageRef, {
      senderId: 'alice',
      recipientId: 'bob',
      textContent: 'Hello future Bob!',
      scheduledFor: futureDate,
      createdAt: new Date(),
      status: 'pending',
    }));
  });
  
  test('should allow recipient to read delivered messages', async () => {
    const alice = testEnv.authenticatedContext('alice');
    const bob = testEnv.authenticatedContext('bob');
    const messageRef = doc(alice.firestore(), 'scheduledMessages', 'message2');
    
    const futureDate = new Date();
    futureDate.setDate(futureDate.getDate() + 1);
    
    // Alice creates message for Bob
    await assertSucceeds(setDoc(messageRef, {
      senderId: 'alice',
      recipientId: 'bob',
      textContent: 'Hello Bob!',
      scheduledFor: futureDate,
      createdAt: new Date(),
      status: 'pending',
    }));
    
    // Bob should be able to read the message
    await assertSucceeds(getDoc(doc(bob.firestore(), 'scheduledMessages', 'message2')));
  });
  
  test('should deny creating messages scheduled for the past', async () => {
    const alice = testEnv.authenticatedContext('alice');
    const messageRef = doc(alice.firestore(), 'scheduledMessages', 'pastMessage');
    
    const pastDate = new Date();
    pastDate.setDate(pastDate.getDate() - 1);
    
    await assertFails(setDoc(messageRef, {
      senderId: 'alice',
      recipientId: 'bob',
      textContent: 'This should fail',
      scheduledFor: pastDate,
      createdAt: new Date(),
      status: 'pending',
    }));
  });
  
  test('should deny messages with empty content', async () => {
    const alice = testEnv.authenticatedContext('alice');
    const messageRef = doc(alice.firestore(), 'scheduledMessages', 'emptyMessage');
    
    const futureDate = new Date();
    futureDate.setDate(futureDate.getDate() + 1);
    
    await assertFails(setDoc(messageRef, {
      senderId: 'alice',
      recipientId: 'bob',
      textContent: '', // Empty content
      scheduledFor: futureDate,
      createdAt: new Date(),
      status: 'pending',
    }));
  });
  
  test('should deny messages exceeding text limit', async () => {
    const alice = testEnv.authenticatedContext('alice');
    const messageRef = doc(alice.firestore(), 'scheduledMessages', 'longMessage');
    
    const futureDate = new Date();
    futureDate.setDate(futureDate.getDate() + 1);
    const longText = 'x'.repeat(5001); // Exceeds 5KB limit
    
    await assertFails(setDoc(messageRef, {
      senderId: 'alice',
      recipientId: 'bob',
      textContent: longText,
      scheduledFor: futureDate,
      createdAt: new Date(),
      status: 'pending',
    }));
  });
  
  test('should allow sender to cancel pending messages', async () => {
    const alice = testEnv.authenticatedContext('alice');
    const messageRef = doc(alice.firestore(), 'scheduledMessages', 'cancelMessage');
    
    const futureDate = new Date();
    futureDate.setDate(futureDate.getDate() + 1);
    
    // Create message
    await assertSucceeds(setDoc(messageRef, {
      senderId: 'alice',
      recipientId: 'bob',
      textContent: 'This will be cancelled',
      scheduledFor: futureDate,
      createdAt: new Date(),
      status: 'pending',
    }));
    
    // Alice should be able to cancel her own message
    await assertSucceeds(updateDoc(messageRef, {
      status: 'cancelled',
    }));
  });
  
  test('should deny non-senders from accessing messages', async () => {
    const alice = testEnv.authenticatedContext('alice');
    const charlie = testEnv.authenticatedContext('charlie');
    const messageRef = doc(alice.firestore(), 'scheduledMessages', 'privateMessage');
    
    const futureDate = new Date();
    futureDate.setDate(futureDate.getDate() + 1);
    
    // Alice creates message for Bob
    await assertSucceeds(setDoc(messageRef, {
      senderId: 'alice',
      recipientId: 'bob',
      textContent: 'Private message',
      scheduledFor: futureDate,
      createdAt: new Date(),
      status: 'pending',
    }));
    
    // Charlie should not be able to read the message
    await assertFails(getDoc(doc(charlie.firestore(), 'scheduledMessages', 'privateMessage')));
  });
});

describe('User Collection Security Rules', () => {
  test('should allow users to read any user profile for search', async () => {
    const alice = testEnv.authenticatedContext('alice');
    const bob = testEnv.authenticatedContext('bob');
    const userRef = doc(alice.firestore(), 'users', 'bob');
    
    // Create Bob's profile
    await assertSucceeds(setDoc(doc(bob.firestore(), 'users', 'bob'), {
      email: 'bob@example.com',
      username: 'bob_user',
      createdAt: new Date(),
    }));
    
    // Alice should be able to read Bob's profile for search
    await assertSucceeds(getDoc(userRef));
  });
  
  test('should only allow users to update their own profile', async () => {
    const alice = testEnv.authenticatedContext('alice');
    const bob = testEnv.authenticatedContext('bob');
    
    // Alice creates her profile
    await assertSucceeds(setDoc(doc(alice.firestore(), 'users', 'alice'), {
      email: 'alice@example.com',
      username: 'alice_user',
      createdAt: new Date(),
    }));
    
    // Alice should be able to update her own profile
    await assertSucceeds(updateDoc(doc(alice.firestore(), 'users', 'alice'), {
      username: 'new_alice_user',
    }));
    
    // Bob should not be able to update Alice's profile
    await assertFails(updateDoc(doc(bob.firestore(), 'users', 'alice'), {
      username: 'hacked_alice',
    }));
  });
  
  test('should validate username format on user creation', async () => {
    const alice = testEnv.authenticatedContext('alice');
    
    // Valid username
    await assertSucceeds(setDoc(doc(alice.firestore(), 'users', 'alice'), {
      email: 'alice@example.com',
      username: 'valid_username123',
      createdAt: new Date(),
    }));
    
    // Invalid username (too short)
    await assertFails(setDoc(doc(alice.firestore(), 'users', 'alice2'), {
      email: 'alice2@example.com',
      username: 'ab',
      createdAt: new Date(),
    }));
    
    // Invalid username (contains special characters)
    await assertFails(setDoc(doc(alice.firestore(), 'users', 'alice3'), {
      email: 'alice3@example.com',
      username: 'user@name',
      createdAt: new Date(),
    }));
  });
});